### ISSUE TITLE
Allow dynamic batch size in the Committer flow 

### ISSUE THREAD

**Issue by sdudzin** [akka/alpakka-kafka#1032]
### Short description

Allow dynamic batch sizes in Committer

### Details

With the release of 2.0.0 the proper way of committing is via Committer API this makes it impossible to control batch sizes individually (as opposed to previous versions where batches were created manually). The proposal is to allow Committer API to take not only `Committable`'s but also prepared batches. 
A use case for this is to allow to commit offsets for successful elements from the batch.

**Comment by ennru** [comment]
`Committable` is the super-type of both `CommittableOffset` and `CommittableOffsetBatch` so you may batch offsets as you like and pass it to the `Committer`.

**Comment by sdudzin** [comment]
`Committer` still uses `settings.maxBatch` to determine fixed the batch size. The only thing that is controllable from 'outside' is `Committable.batchSize` but it's only a 'cost function' for `groupedWeightedWithin` (which still uses  `settings.maxBatch` to determine total size. 

I have attached a PR for review (#1031 ).

**Comment by ennru** [comment]
I'm not sure I understand what you try to achieve.

The `maxBatch` might be misleading indeed, it is more "trigger commit send when at least this many offsets are reported". So it won't split a `CommittableOffsetBatch` into smaller things (eg. if you set `maxBatch` to 1, it will send the whole first thing that comes in to the consumer actor.)

**Comment by sdudzin** [comment]
My use case is to be able to work with batches but in case of stream failure I want to commit the successfully processed part of batch.

For example, I have a batch size of 100 defined by `maxBatch` in case of stream failure in the middle of the batch (for example at element nr 51) I still want to commit batch of first 50 committables and only then fail the stream. So that I wont have to re-read the first 50 duplicates.

With version prior to 2.0.0 it was rather trivial because one could use  commit calls on the batch itself.

**Comment by sdudzin** [comment]
The problem is that the max batch size is baked into the flow definition and thus can not be changed dynamically. With splitting the implementation of the flow (as proposed in the PR) you keep both - the original possibilities and gain ability to dynamically change the batch size.

**Comment by seglo** [comment]
Hi @sdudzin. Last year the offset committing semantics was changed in #862. 

Before, whenever a `Committable` was committed Alpakka Kafka would immediately call `KafkaConsumer.commitAsync`. Default commit configuration meant that in high throughput use cases there could be a lot of in-flight async commit requests at any given time (IIRC the default batch size was `1000`).  We observed that in some situations the commit latency would grow unbounded over time and eventually cause commit timeout failures.  It was decided to change commit semantics so that when a `Committable` is committed it sends a request to the upstream `KafkaConsumerActor`.  The `KafkaConsumerActor` will accumulate all the commit requests and then commit them the next time a poll occurs. Because of this, it's no longer possible to know how many commits are actually committed unless you commit only once per poll interval (50ms by default).

Commit acknowledgements can also cause back-pressure when committing is done in the `Committer.flow|sink` or manually with a `mapAsync`.  Since a commit acknowledgement is only processed during a `KafkaConsumer.poll`, it was possible that when many requests were in-flight that they wouldn't be processed expediently during each poll interval, causing backpressure downstream. With the new changes in #862 there are less potential commit acknowledgements to handle so there's less of a possibility of commits causing back-pressure.  However, we also added an API to let you "commit & forget" in #883, so that it would never be a concern, but this is not used by default.

I realize all of this is complicated, but I think the background is necessary to justify the point I want to make: that committing a batch of any size is only a suggestion.  Commits no longer are sent as soon as a batch criteria is met, so the batch size settings in `CommitSettings`, or the number of records in a custom created `CommittableOffsetBatchImpl` don't really matter anymore. All they enforce is that **at least that batch will be committed at the next poll**.  I hope that makes sense.

**Comment by sdudzin** [comment]
Hi @seglo , thank you for detailed explanation and it makes sense! Our current codebase based on version 1.1.0 which IIRC already contains #862 .

The issue is not really about backpressure or timing of commits to the topic. The issue is that `Committer.batchFlow` uses `groupedWeightedWithin` to build up batches which in turn **only emits when the time elapses or the weight limit reached**. 

Now if the flow gets cancelled half way through - none of that half-built-up batch in the flow will ever be sent to the `consumerActor`. The result is that upon stream restart you would read from last successful offset commit resulting in potentially duplicate messages (at least once strategy).

What we want to _keep doing_ (when upgrading to version 2.0.1+) is being able to catch such stream cancellations, build-up the batch with batch weight maxed-out (obviously with smaller batch size) so that the flow will guaranteedly send the batch to the `consumerActor`  - and only then fail the stream. This allows us to have much more accurate offset commits so that we don't have to re-read messages we know already have been processed.

It's totally fine with batch being sent eventually on the next poll as long as it's sent. Stream cancellation half-way through prevents it from sending partially built up batches at all.

I hope I make more sense?

**Comment by seglo** [comment]
> Now if the flow gets cancelled half way through - none of that half-built-up batch in the flow will ever be sent to the `consumerActor`. The result is that upon stream restart you would read from last successful offset commit resulting in potentially duplicate messages (at least once strategy).
>
> What we want to _keep doing_ (when upgrading to version 2.0.1+) is being able to catch such stream cancellations, build-up the batch with batch weight maxed-out (obviously with smaller batch size) so that the flow will guaranteedly send the batch to the `consumerActor` - and only then fail the stream. This allows us to have much more accurate offset commits so that we don't have to re-read messages we know already have been processed.

I see, but if there's a stream failure how did you commit a partially processed batch?

**Comment by sdudzin** [comment]
In very short version by using `recover` on the flow and then separating successes and the error in sequential (and weighted) groups of messages. After the 'successful' group gets committed you can re-fail the stream with the error from the second group.

**Comment by seglo** [comment]
I see. In the `recover` you call commit on the partial batch you have and that commit request is accumulated in the `KafkaConsumerActor`.  Then when you re-fail the stream any outstanding commit requests are committed before the actor shuts down.

**Comment by sdudzin** [comment]
Very approximately, it's a more sophisticated but fits in a couple dozen lines of code.

**Comment by seglo** [comment]
Have you considered instead of batching at all to just set a `maxBatch` (weight) of 1? Since offset accumulation is happening in the `KafkaConsumerActor` this will ensure that every record that passes through the `Committer.flow` will be attempt to be committed at the enxt poll, or when there's a downstream failure.

**Comment by sdudzin** [comment]
This would imply relying on implicit knowledge and could potentially break with future updates without notice. I'd like to avoid situations like that.

**Comment by seglo** [comment]
I understand, but that's my original point. When commits are actually committed to Kafka are abstracted away from the user now and have been since 1.1.0. Your PR #1031 doesn't change that.

**Comment by sdudzin** [comment]
Indeed it does not change that and in normal scenario's it allows to reduce number of commits by using larger batches and have benefit of 'accurate' offset commits in case of stream failures.

What is needed to be able to achieve such possibilities? What can I do?

**Comment by seglo** [comment]
I would suggest using a smaller `CommitterSettings.maxBatch` as I mentioned earlier. You can also modify the `akka.kafka.consumer.poll-interval` to commit more frequently (the default is `50ms`, though it can occur more frequently when there's no back-pressure).

Most importantly, the system should be able to handle duplicates in some way, because even with direct control over when commits are sent it's still possible that failures occur during the commit process and messages will be processed again when the stream comes back online.  The best we can do is to try to limit how many messages need to be reprocessed at the sacrifice of performance (by committing more frequently).

**Comment by sdudzin** [comment]
@seglo , the system is perfectly capable of dealing with duplicates - this is efficiency and performance optimisation. Unfortunately with smaller batch size we wont achieve such efficiency. 

> You can also modify the akka.kafka.consumer.poll-interval to commit more frequently

The idea is exact opposite - have less frequent commits in normal scenario, and have much less (close to zero) duplicates in case of stream restart due to an error or shutdown due to accurately calculated batch size.

Without these possibilities this means we have a functional regression in the system when upgrading to the latest version and stand before choice of maintaining a fork or looking for alternatives.  

What are the problems with considering the proposed solution? How can I help?

**Comment by seglo** [comment]
@sdudzin Got it.  I think I understand your use case now as you want to optimize your system so that less duplicates need to be reprocessed after a failure.

Instead of the current PR I would like to propose that we implement a new feature in Alpakka Kafka to flush an inflight batch to be committed when an upstream error is encountered.  I think this will benefit both your use case, where you currently have a custom implementation in place, and maintain the current committing abstraction that we want to keep in place.  This would require a rewrite of the batching logic within `Committer`.  We can't use `groupWeightedWithin` on its own, because we don't have access to the inflight batch that we would want to commit when an upstream failure occurs.

Before I think about this in more detail I want to get your feedback about 1) Do I understand the issue correctly?, and 2) Are you open to working on a feature specification together to flush commits on upstream failure?

**Comment by sdudzin** [comment]
@seglo yes you understand my use case correctly, and yes, I am open to help.

**Comment by seglo** [comment]
@sdudzin Great.  I've created a new issue feature request: #1039.  Let's discuss any implementation details there.  Are you OK with closing this issue and the accompanying PR #1031?

**Comment by sdudzin** [comment]
Sure

**Comment by SemanticBeeng** [comment]
>  catch such stream cancellations, build-up the batch with batch weight maxed-out (obviously with smaller batch size) so that the flow will guaranteedly send the batch to the consumerActor - and only then fail the stream

Could this not be done best with `*ManualOffsetSource`s ?
https://github.com/akka/alpakka-kafka/issues/1086#issuecomment-614240326
